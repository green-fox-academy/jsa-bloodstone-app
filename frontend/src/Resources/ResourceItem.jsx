import React from 'react';
import {
  View, Image, StyleSheet, Text,
} from 'react-native';
import PropTypes from 'prop-types';
import AnimateNumber from 'react-native-animate-number';
import Colors from '../common/colors';

import cookieIcon from '../../assets/troop/cookie.png';
import goldIcon from '../../assets/gold.png';

const ICONS = {
  gold: { type: 'gold', icon: goldIcon },
  cookie: { type: 'cookie', icon: cookieIcon },
};

const styles = StyleSheet.create({
  container: {
    paddingVertical: 7,
    paddingHorizontal: 10,
    flexDirection: 'row',
    alignItems: 'center',
  },
  iconStyle: {
    width: 14,
    height: 14,
    marginHorizontal: 4,
  },
  text: {
    fontSize: 14,
    color: Colors.textColor,
    fontWeight: 'bold',
  },
  subText: {
    fontSize: 10,
    color: Colors.greenColor,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});

const getRateText = (rate) => {
  let rateStr;
  let textColor;
  if (rate < 0) {
    rateStr = `${rate}/min`;
    textColor = Colors.orangeColor;
  } else {
    rateStr = `+${rate}/min`;
    textColor = Colors.greenColor;
  }
  return { rateStr, textColor };
};

function handleAmount(amount) {
  if (amount >= 10000000) {
    return `${Number.parseInt(amount / 1000000, 10)}M`;
  }
  if (amount >= 10000) {
    return `${Number.parseInt(amount / 1000, 10)}K`;
  }
  return `${Number.parseInt(amount, 10)}`;
}

function ResourceItem({ type, amount, rate }) {
  const { rateStr, textColor } = getRateText(rate);
  const { icon } = ICONS[type];

  return (
    <View style={styles.container}>
      <View style={{ flexDirection: 'row', alignItems: 'center' }}>
        <View style={styles.row}>
          <Image style={styles.iconStyle} source={icon} />
          <AnimateNumber
            style={styles.text}
            value={amount}
            timing="easeOut"
            formatter={handleAmount}
          />
        </View>
        <View style={{ marginLeft: 4 }}>
          <Text style={{ ...styles.subText, color: textColor }}>{rateStr}</Text>
        </View>
      </View>
    </View>
  );
}

ResourceItem.propTypes = {
  type: PropTypes.oneOf(Object.keys(ICONS)).isRequired,
  amount: PropTypes.number.isRequired,
  rate: PropTypes.number.isRequired,
};

export default ResourceItem;
