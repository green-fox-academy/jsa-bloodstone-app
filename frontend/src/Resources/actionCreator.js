import { SERVER_URL } from 'react-native-dotenv';

export const FETCH_RESOURCES_REQUEST = 'fetchResourcesRequest';
export const FETCH_RESOURCES_SUCCESS = 'fetchResourcesSuccess';
export const FETCH_RESOURCES_FAILURE = 'fetchResourcesFailure';

const URL = `http://${SERVER_URL}/kingdom/resources`;

export function fetchResources(token) {
  return (dispatch) => {
    dispatch({ type: FETCH_RESOURCES_REQUEST });
    return fetch(URL, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
    })
      .then((response) => {
        if (response.status === 200) {
          return response.json();
        }
        throw new Error('An error has occurred, please try later!');
      })
      .then((response) => dispatch({ type: FETCH_RESOURCES_SUCCESS, payload: response.resources }))
      .catch((error) => dispatch({ type: FETCH_RESOURCES_FAILURE, payload: error }));
  };
}
