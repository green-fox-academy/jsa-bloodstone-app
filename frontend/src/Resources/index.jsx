import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import {
  View, Image, Text,
  StyleSheet,
} from 'react-native';
import factoryIcon from '../../assets/buildings/factory.png';
import mineIcon from '../../assets/buildings/mine.png';
import cookieIcon from '../../assets/troop/cookie.png';
import goldIcon from '../../assets/gold.png';
import Colors from '../common/colors';
import { CardView } from '../common/components';
import ErrorPopup from '../ErrorPopup';

import { fetchResources } from './actionCreator';

const styles = StyleSheet.create({
  text: {
    fontSize: 16,
    color: Colors.textColor,
    fontWeight: 'bold',
  },
  subText: {
    fontSize: 12,
    color: Colors.greenColor,
  },
  resourceIcon: {
    width: 14,
    height: 14,
    marginHorizontal: 4,
  },
  buildingIcon: {
    width: 32,
    height: 32,
    marginRight: 4,
  },
  rowFlex: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  rowFlexSpaceAround: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
  },
  resourceView: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});

function ResourceView({
  buildingIcon,
  resourceIcon, amount, changeRate,
}) {
  let changeRateStr;
  let textColor;
  if (changeRate < 0) {
    changeRateStr = `${changeRate}/minute`;
    textColor = Colors.orangeColor;
  } else {
    changeRateStr = `+${changeRate}/minute`;
    textColor = Colors.greenColor;
  }

  return (
    <View style={styles.resourceView}>
      <Image style={styles.buildingIcon} source={buildingIcon} />
      <View>
        <View style={styles.rowFlex}>
          <Text style={styles.text}>{amount}</Text>
          <Image style={styles.resourceIcon} source={resourceIcon} />
        </View>
        <View>
          <Text style={{ ...styles.subText, color: textColor }}>{changeRateStr}</Text>
        </View>
      </View>
    </View>
  );
}

ResourceView.propTypes = {
  resourceIcon: PropTypes.number.isRequired,
  buildingIcon: PropTypes.number.isRequired,
  amount: PropTypes.number.isRequired,
  changeRate: PropTypes.number.isRequired,
};

function Resources() {
  const foodAmount = useSelector((state) => state.resources.foodAmount);
  const foodGeneration = useSelector((state) => state.resources.foodGeneration);

  const goldAmount = useSelector((state) => state.resources.goldAmount);
  const goldGeneration = useSelector((state) => state.resources.goldGeneration);

  const error = useSelector((state) => state.resources.error);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchResources());
    const updateResourcesInterval = setInterval(() => dispatch(fetchResources()), 60000);
    return () => clearInterval(updateResourcesInterval);
  }, []);

  if (error) {
    return (
      <ErrorPopup message={`Oops, ${error.message}`} />
    );
  }

  return (
    <CardView style={styles.rowFlexSpaceAround}>
      <ResourceView
        buildingIcon={factoryIcon}
        resourceIcon={cookieIcon}
        amount={foodAmount}
        changeRate={foodGeneration}
      />
      <ResourceView
        buildingIcon={mineIcon}
        resourceIcon={goldIcon}
        amount={goldAmount}
        changeRate={goldGeneration}
      />
    </CardView>
  );
}

export default Resources;
