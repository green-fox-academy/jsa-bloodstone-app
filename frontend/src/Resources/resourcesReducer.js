import { FETCH_RESOURCES_REQUEST, FETCH_RESOURCES_SUCCESS, FETCH_RESOURCES_FAILURE } from './actionCreator';

const initialState = {
  food: 0,
  foodGeneration: 0,
  gold: 0,
  goldGeneration: 0,
  isLoading: false,
  error: undefined,
};

export default function resources(state = initialState, action) {
  switch (action.type) {
    case FETCH_RESOURCES_REQUEST:
      return { ...state, isLoading: true };
    case FETCH_RESOURCES_SUCCESS:
      return {
        ...state,
        food: action.payload[0].amount,
        foodGeneration: action.payload[0].generation,
        gold: action.payload[1].amount,
        goldGeneration: action.payload[1].generation,
        isLoading: false,
      };
    case FETCH_RESOURCES_FAILURE:
      return { ...state, isLoading: false, error: action.payload };
    default:
      return state;
  }
}
