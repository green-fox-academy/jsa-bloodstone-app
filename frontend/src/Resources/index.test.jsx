import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { useDispatch, useSelector } from 'react-redux';

import Resources from '.';

jest.mock('react-redux');
jest.mock('./actionCreator');

const resources = [
  {
    type: 'food',
    amount: 1578153586,
    generation: 60,
  },
  {
    type: 'gold',
    amount: 1578153704,
    generation: 60,
  },
];

describe('<Resources />', () => {
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => {});
  });

  afterAll(() => {
    jest.resetModules();
  });

  it('Should render resources and match the snapshot', () => {
    const mockedState = {
      resources: {
        foodAmount: resources[0].amount,
        foodGeneration: resources[0].generation,
        goldAmount: resources[1].amount,
        goldGeneration: resources[1].generation,
        isLoading: false,
        error: undefined,
      },
      auth: {
        token: 'test token',
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));

    const renderer = new ShallowRenderer();
    const tree = renderer.render(<Resources />);

    expect(tree).toMatchSnapshot();
  });

  it('Should render loading and match the snapshot', () => {
    const mockedState = {
      resources: {
        foodAmount: resources[0].amount,
        foodGeneration: resources[0].generation,
        goldAmount: resources[1].amount,
        goldGeneration: resources[1].generation,
        isLoading: true,
        error: undefined,
      },
      auth: {
        token: 'test token',
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));

    const renderer = new ShallowRenderer();
    const tree = renderer.render(<Resources />);

    expect(tree).toMatchSnapshot();
  });

  it('Should render error and match the snapshot', () => {
    const mockedState = {
      resources: {
        foodAmount: resources[0].amount,
        foodGeneration: resources[0].generation,
        goldAmount: resources[1].amount,
        goldGeneration: resources[1].generation,
        isLoading: true,
        error: new Error('test error'),
      },
      auth: {
        token: 'test token',
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));

    const renderer = new ShallowRenderer();
    const tree = renderer.render(<Resources />);

    expect(tree).toMatchSnapshot();
  });
});
