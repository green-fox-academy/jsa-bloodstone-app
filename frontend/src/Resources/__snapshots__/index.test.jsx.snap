// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResourceView /> Should render error and match the snapshot 1`] = `
<Text>
  Oops, test error
</Text>
`;

exports[`<ResourceView /> Should render loading and match the snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 6,
        "elevation": 3,
        "marginBottom": 10,
        "padding": 15,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "x": 0,
          "y": 4,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 6,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 32,
          "marginRight": 4,
          "width": 32,
        }
      }
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#55565a",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          1578153586
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 14,
              "marginHorizontal": 4,
              "width": 14,
            }
          }
        />
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#68AA2F",
              "fontSize": 12,
            }
          }
        >
          +60/minute
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 32,
          "marginRight": 4,
          "width": 32,
        }
      }
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#55565a",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          1578153704
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 14,
              "marginHorizontal": 4,
              "width": 14,
            }
          }
        />
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#68AA2F",
              "fontSize": 12,
            }
          }
        >
          +60/minute
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<ResourceView /> Should render resources and match the snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 6,
        "elevation": 3,
        "marginBottom": 10,
        "padding": 15,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "x": 0,
          "y": 4,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 6,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 32,
          "marginRight": 4,
          "width": 32,
        }
      }
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#55565a",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          1578153586
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 14,
              "marginHorizontal": 4,
              "width": 14,
            }
          }
        />
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#68AA2F",
              "fontSize": 12,
            }
          }
        >
          +60/minute
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 32,
          "marginRight": 4,
          "width": 32,
        }
      }
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#55565a",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          1578153704
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 14,
              "marginHorizontal": 4,
              "width": 14,
            }
          }
        />
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#68AA2F",
              "fontSize": 12,
            }
          }
        >
          +60/minute
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Resources /> Should render error and match the snapshot 1`] = `
<Text>
  Oops, test error
</Text>
`;

exports[`<Resources /> Should render loading and match the snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 6,
        "elevation": 3,
        "marginBottom": 10,
        "padding": 15,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "x": 0,
          "y": 4,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 6,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 32,
          "marginRight": 4,
          "width": 32,
        }
      }
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#55565a",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          1578153586
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 14,
              "marginHorizontal": 4,
              "width": 14,
            }
          }
        />
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#68AA2F",
              "fontSize": 12,
            }
          }
        >
          +60/minute
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 32,
          "marginRight": 4,
          "width": 32,
        }
      }
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#55565a",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          1578153704
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 14,
              "marginHorizontal": 4,
              "width": 14,
            }
          }
        />
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#68AA2F",
              "fontSize": 12,
            }
          }
        >
          +60/minute
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Resources /> Should render resources and match the snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 6,
        "elevation": 3,
        "marginBottom": 10,
        "padding": 15,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "x": 0,
          "y": 4,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 6,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 32,
          "marginRight": 4,
          "width": 32,
        }
      }
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#55565a",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          1578153586
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 14,
              "marginHorizontal": 4,
              "width": 14,
            }
          }
        />
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#68AA2F",
              "fontSize": 12,
            }
          }
        >
          +60/minute
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 32,
          "marginRight": 4,
          "width": 32,
        }
      }
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#55565a",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          1578153704
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 14,
              "marginHorizontal": 4,
              "width": 14,
            }
          }
        />
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#68AA2F",
              "fontSize": 12,
            }
          }
        >
          +60/minute
        </Text>
      </View>
    </View>
  </View>
</View>
`;
